gnovm/cmd/gno/mod.go

```diff
@@ -204,13 +206,70 @@ func execModInit(args []string) error {
 	if err != nil {
 		return err
 	}
-	if err := gnomod.CreateGnoModFile(dir, modPath); err != nil {
+	if err := CreateGnoModFile(dir, modPath); err != nil {
 		return fmt.Errorf("create gno.mod file: %w", err)
 	}
 
 	return nil
 }
 
+func CreateGnoModFile(rootDir, modPath string) error {
+	if !filepath.IsAbs(rootDir) {
+		return fmt.Errorf("dir %q is not absolute", rootDir)
+	}
+
+	modFilePath := filepath.Join(rootDir, "gno.mod")
+	if _, err := os.Stat(modFilePath); err == nil {
+		return errors.New("gno.mod file already exists")
+	}
+
+	if modPath == "" {
+		// Check .gno files for package name
+		// and use it as modPath
+		files, err := os.ReadDir(rootDir)
+		if err != nil {
+			return fmt.Errorf("read dir %q: %w", rootDir, err)
+		}
+
+		var pkgName gnolang.Name
+		for _, file := range files {
+			if file.IsDir() || !strings.HasSuffix(file.Name(), ".gno") || strings.HasSuffix(file.Name(), "_filetest.gno") {
+				continue
+			}
+
+			fpath := filepath.Join(rootDir, file.Name())
+			bz, err := os.ReadFile(fpath)
+			if err != nil {
+				return fmt.Errorf("read file %q: %w", fpath, err)
+			}
+
+			pn := gnolang.PackageNameFromFileBody(file.Name(), string(bz))
+			if strings.HasSuffix(string(pkgName), "_test") {
+				pkgName = pkgName[:len(pkgName)-len("_test")]
+			}
+			if pkgName == "" {
+				pkgName = pn
+			}
+			if pkgName != pn {
+				return fmt.Errorf("package name mismatch: [%q] and [%q]", pkgName, pn)
+			}
+		}
+		if pkgName == "" {
+			return errors.New("cannot determine package name")
+		}
+		modPath = string(pkgName)
+	}
+	if err := module.CheckImportPath(modPath); err != nil {
+		return err
+	}
+
+	modfile := new(gnomod.File)
+	modfile.AddModuleStmt(modPath)
+	modfile.Write(filepath.Join(rootDir, "gno.mod"))
```

Comment:
unhandled errors...

Reply:
> It's not introduced in this PR, I just moved this function to a new file in this PR.
I will open another PR soon to divide it into smaller parts.

I will also fix it in the same PR